source/extensions/http/cache/file_system_http_cache/cache_file_header_proto_util.cc

```diff
@@ -0,0 +1,96 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_file_header_proto_util.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+CacheFileHeader mergeProtoWithHeadersAndMetadata(const CacheFileHeader& entry_headers,
+                                                 const Http::ResponseHeaderMap& response_headers,
+                                                 const ResponseMetadata& response_metadata) {
+  Http::ResponseHeaderMapPtr merge_headers = headersFromHeaderProto(entry_headers);
+  applyHeaderUpdate(response_headers, *merge_headers);
+  return makeCacheFileHeaderProto(entry_headers.key(), *merge_headers, response_metadata);
+}
+
+CacheFileHeader makeCacheFileHeaderProto(const Key& key,
+                                         const Http::ResponseHeaderMap& response_headers,
+                                         const ResponseMetadata& metadata) {
+  CacheFileHeader file_header;
+  *file_header.mutable_key() = key;
+  TimestampUtil::systemClockToTimestamp(metadata.response_time_,
+                                        *file_header.mutable_metadata_response_time());
+  response_headers.iterate([&file_header](const Http::HeaderEntry& header) {
+    CacheFileHeader::Header* h = file_header.add_headers();
+    h->set_key(std::string{header.key().getStringView()});
+    h->set_value(std::string{header.value().getStringView()});
+    return Http::HeaderMap::Iterate::Continue;
+  });
+  return file_header;
+}
+
+CacheFileTrailer makeCacheFileTrailerProto(const Http::ResponseTrailerMap& response_trailers) {
+  CacheFileTrailer file_trailer;
+  response_trailers.iterate([&file_trailer](const Http::HeaderEntry& trailer) {
+    CacheFileTrailer::Trailer* t = file_trailer.add_trailers();
+    t->set_key(std::string{trailer.key().getStringView()});
+    t->set_value(std::string{trailer.value().getStringView()});
+    return Http::HeaderMap::Iterate::Continue;
+  });
+  return file_trailer;
+}
+
+size_t headerProtoSize(const CacheFileHeader& proto) { return proto.ByteSizeLong(); }
+
+Buffer::OwnedImpl bufferFromProto(const CacheFileHeader& proto) {
+  // TODO(ravenblack): consider proto.SerializeToOStream and an ostream impl to Buffer.
```

Comment:
or zero-copy.

Reply:
Changed to zero copy since it's clearly nicer than ostream.